[#179506968] Enable point-in-time restore for MySQL

AP-Hunt opened the pull request:
What
---
The feature is available and used in production, but doesn't have an
integration test.

How to test
---

~Don't, for now.~ See further down :point_down:

AP-Hunt committed:
Add a test for point-in-time restores for Postgres 12

The feature is available and used in production, but doesn't have an
integration test.

risicle committed:
sqlengine: implement MySQLEngine.ResetState()

risicle committed:
rdsbroker: enable restoreFromPointInTime and restoreFromSnapshot for mysql

risicle committed:
integration tests: enable mysql restore and PITR tests

risicle committed:
integration tests: reduce concurrency

as it makes RDS unhappy with our api request rate

risicle committed:
README.md: add some notes on integration tests

AP-Hunt commented on sqlengine/mysql_engine.go:
```diff
@@ -141,8 +140,52 @@ func (d *MySQLEngine) DropUser(bindingID string) error {
 func (d *MySQLEngine) ResetState() error {
-	// TODO: Not implemented
-	return errors.New("Not implemented")
+	d.logger.Debug("reset-state.start")
```
The Lager way of doing this is

```
logSession := d.logger.Session("reset-state")
logSession.Debug("start")
```

That produces a message/action like `things.coming.before.it.reset-state.start`

We don't usually use the Debug level either, because we want these messages to appear in our logs so we can debug it later down the line. We typically use the Info level.

risicle replied:
Ok I just copy-pasta'd this from the postgres implementation, like much of it TBH.

risicle replied:
Urgh, looking into this, making it work this way would be out of kilter with the way logging is done everywhere else in the broker, and most other methods don't log entry and exit, so I might just remove it. And someone can look at overhauling the log messages in the future if that's what they want.

AP-Hunt commented on rdsbroker/broker.go:
```diff
@@ -260,8 +260,13 @@ func (b *RDSBroker) restoreFromPointInTime(
+		switch engine := *servicePlan.RDSProperties.Engine; engine {
+			case "postgres":
+			case "mysql":
+			default:
```
This statement is confusing at first, and reads a bit like we don't support it for either. Consider

```
if servicePlan.RDSProperties.Engine != nil {
    if servicePlan.RDSProperties.Engine != "postgres" && servicePlan.RDSProperties.Engine != "mysql {
        ....
    }
}

```

Otherwise maybe we could look at adding some methods to the `ServicePlan` type? That could make some of the code a little easier to comprehend

```
func (sp *ServicePlan) SupportsPointInTimeRestore() bool {
    return sp.RDSProperties.Engine == "mysql" || sp.RDSProperties.Engine == "postgres"
}

...

func (b *RDSBroker) restoreFromPointInTime(
	ctx context.Context,
	instanceID string,
	details brokerapi.ProvisionDetails,
	asyncAllowed bool,
	provisionParameters ProvisionParameters,
	servicePlan ServicePlan,
) error {
    if !servicePlan.SupportsPointInTimeRestore() {
        return fmt.Errorf("Restore from snapshot not supported for engine '%s'", servicePlan.RDSPropertoes.Engine)
    } 
}
```

risicle replied:
Ok - you can tell I really just wanted to write `if sp.RDSProperties.Engine in ("mysql", "postgres")` but couldn't :(

AP-Hunt commented on rdsbroker/broker.go:
```diff
@@ -326,8 +331,13 @@ func (b *RDSBroker) restoreFromSnapshot(
-	if servicePlan.RDSProperties.Engine != nil && *servicePlan.RDSProperties.Engine != "postgres" {
-		return fmt.Errorf("Restore from snapshot not supported for engine '%s'", *servicePlan.RDSProperties.Engine)
+	if servicePlan.RDSProperties.Engine != nil {
+		switch engine := *servicePlan.RDSProperties.Engine; engine {
```
Same consideration as for point in time restore code

AP-Hunt commented on README.md:
````diff
@@ -134,6 +134,8 @@ You can run them with:
 make integration
 ```
 
+It's best to resist the temptation to raise the tests' parallelism too far as AWS will throttle too many RDS API interactions for an account.
````
This is a very useful note

AP-Hunt commented on ci/blackbox/rdsbroker_test.go:
```diff
@@ -696,6 +696,10 @@ var _ = Describe("RDS Broker Daemon", func() {
 			TestRestoreFromPointInTime("postgres", "postgres-micro-12")
 		})
+
+		Describe("MySQL 8.0", func() {
```
Why no MySQL 5.7 here?

risicle replied:
Good question. I think I saw only one pg version being tested and thought I'd do the same for mysql, assuming it was to limit the runtime.

risicle replied:
Added it for now so we can at least prove it's working, may remove before merge if it's too slow.

risicle committed:
sqlengine tests: add unit test of MySQLEngine.ResetState()

heavily based on PostgresEngine's

risicle committed:
bump all golang dependencies

this is principally to give us a new mysql driver which doesn't have
sporadic authentication plugin issues with mysql 8.0, but most of our
dependencies are old as hell anyway, so they will likely benefit from
this

tommrr reviewed (approved):
LGTM

AP-Hunt commented on sqlengine/mysql_engine_test.go:
```diff
@@ -136,4 +140,64 @@ var _ = Describe("MySQLEngine", func() {
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("ResetState() removes the credentials", func() {
```
If these tests are required to run in this order, we should merge them in to one test and use `By()` to document them. Ginkgo can run tests in random order, so we might see random failures.

```
It("ResetState() removes the credentials and subsequent CreateUser() returns the same user and different password", func() {
  By("Restting state")
  connectionString := fmt.Sprintf("%s:%s@tcp(%s:%d)/%s", createdUser, createdPassword, address, port, dbname)
  db, err := sql.Open("mysql", connectionString)
  defer db.Close()
  Expect(err).ToNot(HaveOccurred())
  err = db.Ping()
  Expect(err).To(HaveOccurred())
  
  myErr, ok := err.(*mysql.MySQLError)
  Expect(ok).To(BeTrue())
  Expect(myErr.Number).To(SatisfyAny(
	  BeEquivalentTo(1045),
  ))
  Expect(myErr.Message).To(SatisfyAny(
	  MatchRegexp("Access denied.*"),
  ))
  
  By("and then calling CreateUser()")
  user, password, err := mysqlEngine.CreateUser(bindingID, dbname, false)
  Expect(err).ToNot(HaveOccurred())
  Expect(user).To(Equal(createdUser))
  Expect(password).ToNot(Equal(createdPassword))
})
```

risicle replied:
They don't have to run a specific order, but I know why it appears that way - it confused me too at first (again I just copied them from the postgres ones). The function under test, `ResetState()` in this case, actually gets run in the `BeforeEach` clause. So the two `It` clauses are actually totally independent of each other.

I think the confusing thing is actually just the language it uses in the descriptions. If I switched the tenses it would be clearer that the `It` clauses are really just performing assertions.

risicle replied:
Ugh actually fiddling with the language made it more confusing.

AP-Hunt replied:
I get you. Fine with leaving it as is if changing it would make it worse.

AP-Hunt commented on rdsbroker/broker.go:
```diff
@@ -260,8 +260,10 @@ func (b *RDSBroker) restoreFromPointInTime(
 ) error {
-	if servicePlan.RDSProperties.Engine != nil && *servicePlan.RDSProperties.Engine != "postgres" {
-		return fmt.Errorf("Restore from snapshot not supported for engine '%s'", *servicePlan.RDSProperties.Engine)
+	if engine := servicePlan.RDSProperties.Engine; engine != nil {
```
This is cleaner, I like it.

AP-Hunt commented on go.mod:
```diff
@@ -5,22 +5,23 @@ go 1.13
 require (
```
Yay for doing dependency bumps in a separate commit

cadmiumcat reviewed (commented):
I can't run the tests (because I don't have the right permissions) but the code lgtm
